src/ir/possible-contents.cpp

```diff
@@ -641,9 +642,15 @@ struct InfoCollector
     addRoot(curr);
   }
   void visitRefFunc(RefFunc* curr) {
-    addRoot(curr,
-            PossibleContents::literal(
-              Literal::makeFunc(curr->func, curr->type.getHeapType())));
+    if (!getModule()->getFunction(curr->func)->imported()) {
+      // This is not imported, so we know the exact function literal.
+      addRoot(
+        curr,
+        PossibleContents::literal(Literal::makeFunc(curr->func, *getModule())));
+    } else {
+      // This is imported, so it might be anything of the proper type.
+      addRoot(curr);
```

Comment:
Don't we still want to track that this is a reference to the imported function? It would just have an inexact type.

Reply:
I did say "imagine" :laughing: 

But, while we don't have `ref.eq` on functions in wasm userspace, we do have optimizations that compare functions in other ways. E.g. folding an if with `ref.eq` arms, or GUFA inferences. I admit I don't see an actual bug atm in our optimizer, but a future one is conceivable.

Reply:
I can see how optimizations might see that two references are the same and e.g. merge two equivalent `if` arms or something like that, but I still don't see how we could ever have an optimization that does something unsafe when reasoning that two different function references are _different_. Can we at least consider changes here in a separate PR?

Reply:
We can certainly consider changes, sure. If this is a major benefit then maybe it is worth the risk.

But some things that worry me:

* JS can compare exported wasm funcrefs, even if wasm can't. Maybe we integrate wasm with that some day?
* Imagine we have a class of functions that return integer IDs. We know each such function returns a different ID - say this is some property we add with a flag, or some host API works that way which we are aware of (e.g. the Binaryen C API's constant-reporting methods do - this is not entirely hypothetical). Then `f() != g()` can be inferred to be `1` for `f != g`, *unless* they are imports.

None of this seems *immediately* relevant... but it is unsettling.

Reply:
Yes, agreed. I'm not opposed to being conservative here to be on the safe side. But we should keep the less-conservative status quo in this PR to make sure we're not unexpectedly regressing optimizations due to just the introduction of inexact imported functions.

Reply:
What if GUFA has a function literal, but its type isn't exact?

Reply:
I don't see the problem. Once we have exact imports, then the Literal for the imported function would have an exact type iff the import is exact. GUFA would then look at the literal type to see whether casts would succeed or fail, for example. The changes to support inexact function literals are already in this PR.